ckb_type_id.h

```
@@ -0,0 +1,150 @@
+  mol_seg_t args_bytes_seg = MolReader_Bytes_raw_bytes(&args_seg);
+  if (offset + 32 > args_bytes_seg.size) {
+    DEBUG("Length of type id is incorrect!");
+    return CKB_USER_ERROR;
+  }
+  memcpy(type_id, &args_bytes_seg.ptr[offset], 32);
+  return CKB_SUCCESS;
+}
+
+// 1 for true, 0 for false
+int _ckb_has_type_id_cell(size_t index, int is_input) {
+  uint64_t len = 0;
+  size_t source =
+      is_input == 1 ? CKB_SOURCE_GROUP_INPUT : CKB_SOURCE_GROUP_OUTPUT;
+  int ret = ckb_load_cell(NULL, &len, 0, index, source);
+  return ret == CKB_SUCCESS ? 1 : 0;
+}
+
+int _ckb_locate_first_type_id_output_index(uint64_t* index) {
+  uint8_t current_script_hash[32];
+  uint64_t len = 32;
+  int ret = ckb_load_script_hash(current_script_hash, &len, 0);
+  if (ret != CKB_SUCCESS) {
+    DEBUG("Error loading current script hash!");
+    return ret;
+  }
+  if (len != 32) {
+    DEBUG("Invalid script hash length!");
+    return CKB_USER_ERROR;
+  }
+
+  uint64_t i = 0;
+  while (1) {
+    uint8_t buffer[32];
+    len = 32;
+    ret = ckb_load_cell_by_field(buffer, &len, 0, i, CKB_SOURCE_OUTPUT,
+                                 CKB_CELL_FIELD_TYPE_HASH);
+    if (ret != CKB_SUCCESS) {
+      DEBUG("Error fetching output type hash to locate type id index!");
+      return ret;
+    }
+    if (len != 32) {
+      DEBUG("Invalid type hash length!");
+      return CKB_USER_ERROR;
+    }
+    if (memcmp(buffer, current_script_hash, 32) == 0) {
+      break;
+    }
+    i += 1;
+  }
+  *index = i;
+  return CKB_SUCCESS;
+}
+
+int ckb_validate_type_id(size_t offset) {
+  int has_second_type_id_input = _ckb_has_type_id_cell(1, 1);
+  int has_second_type_id_output = _ckb_has_type_id_cell(1, 0);
+  if (has_second_type_id_input || has_second_type_id_output) {
+    DEBUG("There can only be at most one input and one output type ID cell!");
+    return CKB_USER_ERROR;
+  }
+
+  int has_first_type_id_input = _ckb_has_type_id_cell(0, 1);
+  if (!has_first_type_id_input) {
```

missing "else" part. Do we need to return any error code in "else"?

---

We don't need any checks in the else part, it is safe to ignore it